### ISSUE TITLE
Documentation for Agda packages?

### ISSUE THREAD

**Issue by rnhmjoj** [NixOS/nixpkgs#19434]
I have been trying for two days, without success, to load and use agda-stdlib.
The [agda documentation](http://agda.readthedocs.io/en/latest/tools/package-system.html) says a `standard-library.agda-lib` should be found in the directory but it's not the case for agda-stdlib. The file is in the project root but it seems only `/src` is in the nix store.

Also do you really have to manually specify libraries in a file to use them? Can't this be automated?

**Comment by copumpkin** [comment]
I think @Fuuzetsu might have started that stuff, but I haven't tried touching it in a long time.

**Comment by Fuuzetsu** [comment]
@rnhmjoj and anyone who may be concerned in the future:

long ago I have implemented `agda` builder in nixpkgs; multiple libraries that existed at the time were also package; I do not know the state that is in but simply look for `agda` in `all-packages.nix`. With that, you write your package as `agda.mkDerivation` and depend on any of the libraries there. Then in your environment, all the packages you wanted (including stdlib) would be in AGDA_PACKAGE_PATH env var (already compiled).

Having said that, I have no idea what state it is in. Also the linked package manager thing above was not around at all when I wrote it: in fact nix was the first package manager which actually packaged Agda things AFAICT. Looking through the link, it makes me think that setting `agda2-include-dirs` to the value of the variable will no longer work. It may well need updating if you're interested in using it. It seems to me you're trying to use `agda-stdlib` from nixpkgs with the Agda package manager which was never intended.

**Comment by pierrebeaucamp** [comment]
Just for reference:

If you're developing a library, you might be able to use the agda builder. My `default.nix` looks something like:

```nix
{ pkgs ? import <nixpkgs>{} } :

pkgs.agda.mkDerivation (self: rec {
  name = "MyPackage";
  src = ./.;
  sourceDirectories = [ "." ];
  everythingFile = "MyPackage.agda";
  buildDepends = [ pkgs.AgdaStdlib ];
})
```
At least, `nix-build` will spit out some agda related stuff and doesn't report any error.

I'm still figuring out how to properly develop an executable. So far, my approach is to not build with nix at all and rather use an `.agda-lib` file, which looks something like this:

```
name: MyPackage
include: ./.
  , /nix/store/prnkg0xv0575svh8rs5kdlczfwdhv9i1-agda-stdlib-0.13/share/agda 
```

But I guess you might already know this based on your initial comment
>do you really have to manually specify libraries in a file to use them? 

This way you can actually pass some more flags to the agda compiler, which allows you to compile your code into an executable. At least, this works if compiling to javascript. If you decide to compile using GHC, I'm seeing a whole lot of other problems regarding missing GHC libraries.

I'm still pretty new to Agda and Nix and neither of them seem particular UX friendly to me. I'm posting this mainly so other people might discover and/or learn more about the issue, since there is little to no literature regarding Nix & Agda.

**Comment by rnhmjoj** [comment]
Thank you.
I wasn't trying to do anything complex, just playing around in Agda while learning dependent types.
I gave up pretty quickly because as you say the UX is quite bad (especially if you don't use emacs).

**Comment by siddharthist** [comment]
The following file has a (non-minimal) working example of how to build executables. I've even gotten multithreaded GHC to work.

https://github.com/siddharthist/write-yourself-a-scheme-in-agda/blob/master/default.nix

**Comment by razvan-panda** [comment]
@srghma I tried using your script but it did not work. Is anything else required except running it via nix-shell?

**Comment by turion** [comment]
@rnhmjoj This is probably resolved on the latest master with #76653. Can you revisit?

**Comment by rnhmjoj** [comment]
I haven't been paying attention to Agda at all, thank you for linking the PR.
It seems the situation just got much better. With `agda.withPackages` this issue can be closed.
